From 80f3b9c3935baba1bbf2ba730a30f9364105ace8 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Wed, 20 Jan 2016 14:04:08 +0800
Subject: oleaut32: Fix logic for deciding whether type description follows the
 name.

This makes it possible to load an SLTG typelib generated by widl.
---
 dlls/oleaut32/typelib.c | 28 ++++++++--------------------
 1 file changed, 8 insertions(+), 20 deletions(-)

diff --git a/dlls/oleaut32/typelib.c b/dlls/oleaut32/typelib.c
index 0a83b79..e5d8708 100644
--- a/dlls/oleaut32/typelib.c
+++ b/dlls/oleaut32/typelib.c
@@ -4216,35 +4216,23 @@ static void SLTG_DoFuncs(char *pBlk, char *pFirstItem, ITypeInfoImpl *pTI,
 	for(param = 0; param < pFuncDesc->funcdesc.cParams; param++) {
 	    char *paramName = pNameTable + *pArg;
 	    BOOL HaveOffs;
-	    /* If arg type follows then paramName points to the 2nd
-	       letter of the name, else the next WORD is an offset to
-	       the arg type and paramName points to the first letter.
-	       So let's take one char off paramName and see if we're
-	       pointing at an alpha-numeric char.  However if *pArg is
-	       0xffff or 0xfffe then the param has no name, the former
-	       meaning that the next WORD is the type, the latter
-	       meaning that the next WORD is an offset to the type. */
-
-	    HaveOffs = FALSE;
-	    if(*pArg == 0xffff)
-	        paramName = NULL;
-	    else if(*pArg == 0xfffe) {
-	        paramName = NULL;
-		HaveOffs = TRUE;
-	    }
-	    else if(paramName[-1] && !isalnum(paramName[-1]))
-	        HaveOffs = TRUE;
+
+            TRACE_(typelib)("param %d: paramName %s, pArg %p, *pArg %#x\n",
+                param, debugstr_a(paramName), pArg, *pArg);
 
 	    pArg++;
 
+            if (*pArg & 0xff00 || (WORD*)(pBlk + *pArg) < pArg)
+                HaveOffs = FALSE; /* type follows */
+            else
+                HaveOffs = TRUE;
+
 	    if(HaveOffs) { /* the next word is an offset to type */
 	        pType = (WORD*)(pBlk + *pArg);
 		SLTG_DoElem(pType, pBlk,
 			    &pFuncDesc->funcdesc.lprgelemdescParam[param], ref_lookup);
 		pArg++;
 	    } else {
-		if(paramName)
-		  paramName--;
 		pArg = SLTG_DoElem(pArg, pBlk,
                                    &pFuncDesc->funcdesc.lprgelemdescParam[param], ref_lookup);
 	    }
-- 
2.6.4

